sys/usb/usbus/msc/scsi.c

```diff
@@ -0,0 +1,386 @@
+
+    /* Clear the exact amount of bytes needed for this transfer */
+    memset(pkt, 0, len);
+
+    /* Compute the MTD sector size */
+    mtd_sector_size = mtd->page_size * mtd->pages_per_sector;
+
+    /* Check mtd sector size boundaries */
+    /* Host side expects to have at least 512 bytes per block, otherwise it
+     * will reject the setup. Thus, add some logic here to be able to use
+     * smaller page size mtd device */
+    if (mtd_sector_size < 512) {
+        block_size = 512;
+        /* Compute the number of block that will be reported to the host
+           Since the MTD sector is less than the minimum allowed (512),
+           the number of block reported is divided by the number of
+           sector that will be written
+           Example: if the MTD sector size is 64 bytes and the USB reported
+           block is 512, we must use 8 (512 / 64) sectors per USB block */
+        blk_nb /= (512 / mtd_sector_size);
+    }
+    /* setup some logic to manage devices with more than 4096 bytes
+     * per page size */
+    else if (mtd_sector_size > 4096) {
+        /* Force it to the greatest value supported by host */
+        block_size = 4096;
+        /* This case is the reverse as above, since the size of a sector
+           is above the maximum allowed boundary, Read/Modify/Write
+           operation is needed to properly write to the MTD device */
+           blk_nb *= (mtd_sector_size / 4096);
+    }
+    /* Standard page size support by Linux host driver (see drivers/scsi/sd.c
+     * in Linux kernel source tree) */
+    else if (mtd_sector_size == 512  || mtd_sector_size == 1024 ||
+             mtd_sector_size == 2048 || mtd_sector_size == 4096) {
+        block_size = mtd_sector_size;
+    }
+    else {
+        DEBUG_PUTS("Unsupported pagesize");
+        assert(0);
+    }
+    /* Report the size of the block (mtd sector size) */
+    pkt->blk_len.u32 = byteorder_swapl(block_size);
+    /* Store the reported block size and number for internal use */
+    msc->lun_dev[lun].block_size = block_size;
+    msc->lun_dev[lun].block_nb = blk_nb;
+    /* Report the last block number */
+    pkt->last_blk.u32 = byteorder_swapl(blk_nb - 1);
+
+    /* copy into ep buffer */
+    usbdev_ep_xmit(msc->ep_in->ep, (uint8_t *)pkt, len);
+    msc->state = WAIT_FOR_TRANSFER;
+}
+
+static void _scsi_request_sense(usbus_handler_t *handler)
+{
+    usbus_msc_device_t *msc = container_of(handler, usbus_msc_device_t,
+                                           handler_ctrl);
+    msc_request_sense_pkt_t *pkt = (msc_request_sense_pkt_t *)msc->in_buf;
+    size_t len = sizeof(msc_request_sense_pkt_t);
+
+    /* Clear the exact amount of bytes needed for this transfer */
+    memset(pkt, 0, len);
+
+    pkt->error_code = SCSI_REQUEST_SENSE_ERROR;
+    pkt->sense_key = SCSI_SENSE_KEY_ILLEGAL_REQUEST;
+    /* Report the size of the struct minus the 7 first bytes */
+    pkt->add_len = len - 7;
+    pkt->asc = 0x30;
+    pkt->ascq = 0x01;
```

Comment:
What are those magic numbers?

Reply:
Is undocumented black magic numbers, a valid answer here ?

Reply:
Where did you get them from?

Reply:
Hmm, I think it was on some source code provided by ST for a STM32F1 that I used back in 2019.
I did have a look at Zephyr a few months ago too, but it was the same. (Just hexadecimal value used to fill their buffer).
I'll have a look at it tonight. I have a bunch of USB MSC related document on my home laptop. I'll see if I can give you more insight then.